zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -246,18 +246,23 @@ void request(Request request) throws IOException {
             LOG.error("Throttled request sent to leader: {}. Exiting", request);
             ServiceUtils.requestSystemExit(ExitCode.UNEXPECTED_ERROR.getValue());
         }
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        DataOutputStream oa = new DataOutputStream(baos);
-        oa.writeLong(request.sessionId);
-        oa.writeInt(request.cxid);
-        oa.writeInt(request.type);
-        if (request.request != null) {
+        // size of sessionId, cxId and type in bytes
+        int size = Long.BYTES + 2 * Integer.BYTES;
+        byte[] bytes = null;
+        if(request.request != null){
             request.request.rewind();
             int len = request.request.remaining();
             byte[] b = new byte[len];
             request.request.get(b);
-            request.request.rewind();
-            oa.write(b);
+            size = size + len;
+        }
+        ByteArrayOutputStream baos = new ByteArrayOutputStream(size);
+        DataOutputStream oa = new DataOutputStream(baos);
+        oa.writeLong(request.sessionId);
+        oa.writeInt(request.cxid);
+        oa.writeInt(request.type);
+        if (bytes != null) {
```

Comment:
in your code `bytes` is always `null` now, you never set it.

Reply:
I thought it was only related with this method. But I can go through all the places where ByteArrayOutputStream  is used. I will also go through other classes. Thank you for your feedback :)